LegendaryExplorer/LegendaryExplorer/Tools/Dialogue Editor/DialogueEditorExperiments/DialogueEditorExperimentsE.cs

```diff
@@ -132,19 +115,128 @@ private static void updateFaceFX(FaceFXAnimSetEditorControl fxa, string oldRef,
         }
 
-        private static string promptForRef(string msg, string err)
-        {
-            if (PromptDialog.Prompt(null, msg) is string stringRef)
-            {
+        private static string promptForRef(string msg, string err) {
+            if (PromptDialog.Prompt(null, msg) is string stringRef) {
                 int intRef;
-                if (string.IsNullOrEmpty(stringRef) || !int.TryParse(stringRef, out intRef))
-                {
+                if (string.IsNullOrEmpty(stringRef) || !int.TryParse(stringRef, out intRef)) {
                     MessageBox.Show(err, "Warning", MessageBoxButton.OK);
                     return null;
                 }
                 return intRef.ToString();
             }
             return null;
         }
+        #endregion
+
+        #region Clone Node And Sequence
+        /// <summary>
+        /// Clones a Dialogue Node and its related Sequence, while giving it a unique id.
+        /// </summary>
+        /// <param name="dew">Dialogue Editor Window instance.</param>
+        public static void CloneNodeAndSequence(DialogueEditorWindow dew) {
+            DialogueNodeExtended selectedDialogueNode = dew.SelectedDialogueNode;
+
+            if (dew.Pcc != null && selectedDialogueNode != null) {
+                // Need to check if the node has associated data
+                if (selectedDialogueNode.Interpdata == null) {
+                    MessageBox.Show("The selected node does not have an InterpData associated with it.", "Warning", MessageBoxButton.OK);
+                    return;
+                }
+
+                int newID = promptForID("New node ExportID:", "Not a valid ExportID.");
+                if (newID == 0) { return; }
+
+                if (selectedDialogueNode.ExportID.Equals(newID)) {
+                    MessageBox.Show("New ExportID matches the existing one.", "Warning", MessageBoxButton.OK);
+                    return;
+                }
+
+                ExportEntry oldInterpData = selectedDialogueNode.Interpdata;
+
+                // Get the Interp linked to the InterpData
+                IEnumerable<KeyValuePair<IEntry, List<string>>> interpDataReferences = oldInterpData.GetEntriesThatReferenceThisOne()
+                    .Where(entry => entry.Key.ClassName == "SeqAct_Interp");
+                if (interpDataReferences.Count() > 1) {
+                    MessageBox.Show("The selected Node's InterpData is references by more than one Interp. Please ensure it's only references by one.", "Warning", MessageBoxButton.OK);
+                }
+                ExportEntry oldInterp = (ExportEntry)interpDataReferences.First().Key;
+
+                // Get the/a ConvNode linked to the Interp
+                ExportEntry oldConvNode = SeqTools.FindOutboundConnectionsToNode(oldInterp, SeqTools.GetAllSequenceElements(oldInterp).OfType<ExportEntry>())
+                    .Where(entry => entry.ClassName == "BioSeqEvt_ConvNode").ToList().First();
```

Comment:
Instead of `.Where(entry => entry.ClassName == "BioSeqEvt_ConvNode").ToList().First()`, you can just do `.First(entry => entry.ClassName == "BioSeqEvt_ConvNode")` (see the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.first?view=net-6.0#system-linq-enumerable-first-1(system-collections-generic-ienumerable((-0))-system-func((-0-system-boolean))))). 

I would encourage you to use `FirstOrDefault` and then check for null instead of using `First` though. If there is no element that satisfies the condition, `First` will throw an exception, whereas `FirstOrDefault` will return the default value for the type (null in this case). This allows you to handle the lack of the element more gracefully. You should only use `First` when you are absolutely certain that it will work.